src/misc/VarNames.jl

```diff
@@ -0,0 +1,436 @@
+
+julia> R, vec = polynomial_ring(ZZ, AbstractAlgebra.variable_names(s...))
+(Multivariate polynomial ring in 7 variables over integers, AbstractAlgebra.Generic.MPoly{BigInt}[a, b, x11, x12, y1, y2, z])
+
+julia> (a, b), x, y, z = AbstractAlgebra.reshape_to_varnames(vec, s...)
+(AbstractAlgebra.Generic.MPoly{BigInt}[a, b], AbstractAlgebra.Generic.MPoly{BigInt}[x11 x12], AbstractAlgebra.Generic.MPoly{BigInt}[y1, y2], z)
+
+julia> R, (a, b), x, y, z = polynomial_ring(ZZ, s...)
+(Multivariate polynomial ring in 7 variables over integers, AbstractAlgebra.Generic.MPoly{BigInt}[a, b], AbstractAlgebra.Generic.MPoly{BigInt}[x11 x12], AbstractAlgebra.Generic.MPoly{BigInt}[y1, y2], z)
+
+```
+"""
+reshape_to_varnames(vec::Vector, varnames::VarNames...) =
+    reshape_to_varnames(vec, varnames)
+function reshape_to_varnames(vec::Vector, varnames::Tuple{Vararg{VarNames}})
+    iter = Iterators.Stateful(vec)
+    result = Tuple(_reshape_to_varnames(iter, x) for x in varnames)
+    @assert isempty(iter)
+    return result
+end
+
+_reshape_to_varnames(iter::Iterators.Stateful, ::VarName) = popfirst!(iter)
+_reshape_to_varnames(iter::Iterators.Stateful, a::AbstractArray{<:VarName}) =
+    _reshape(iter, size(a))
+_reshape_to_varnames(iter::Iterators.Stateful, (_, shape)::Pair{<:VarName}) =
+    __reshape(iter, shape)
+
+__reshape(iter, axes::Tuple) = _reshape(iter, Int[d for axe in axes for d in size(axe)])
+__reshape(iter, axe) = _reshape(iter, size(axe))
+
+_reshape(iter, dims) = reshape(collect(Iterators.take(iter, prod(dims))), Tuple(dims))
+
+"""
+    keyword_arguments((kvs::Expr...), default::Dict, [valid::Dict]) :: Dict
+
+Mimic usual keyword arguments for usage in macros.
+
+* `kvs`: tuple of Expr of form :(k = v)
+* `default`: dictionary providing the allowed keys and their default values
+* `valid`: optional `Dict{Symbol, <:AbstractVector}` constraining the valid values for some keys
+
+Return a copy of `default` with the key value pairs from `kvs` applied.
+
+# Example
+```jldoctest; setup = :(using AbstractAlgebra)
+julia> AbstractAlgebra.keyword_arguments((:(a=1), :(b=:yes)),
+       Dict(:a=>0, :b=>:no, :c=>0),
+       Dict(:b => [:(:yes), :(:no)]))
+Dict{Symbol, Any} with 3 entries:
+  :a => 1
+  :b => :(:yes)
+  :c => 0
+```
+"""
+function keyword_arguments(kvs::Tuple{Vararg{Expr}}, default::Dict{Symbol},
+        valid::Dict{Symbol, <:AbstractVector} = Dict{Symbol, Vector{Any}}()) ::
+        Dict{Symbol}
+    result = Dict{Symbol, Any}(default)
+    for o in kvs
+        req(MT.@capture(o, k_ = v_), "Only key value options allowed")
+        req(k in keys(result), "Invalid key value option key `$k`")
+        k in keys(valid) && req(v in valid[k], "Invalid option `$v` to key `$k`")
+        result[k] = v
+    end
+    return result
+end
+
+function _eval(m::Core.Module, e::Expr)
+    try
+        Base.eval(m, e)
+    catch err
+        if isa(err, UndefVarError)
+            @error "Inconveniently, you may only use literals and variables " *
```

Comment:
```suggestion
            @error "You may only use literals and variables " *
```

Reply:
Oh, but I always smile, when reading that error message 😄